src/accountsdb/download.zig

```diff
@@ -0,0 +1,652 @@
+                    peer.full_snapshot.slot,
+                    inc_snapshot.slot,
+                    inc_snapshot.hash,
+                    "tar.zst",
+                });
+                defer allocator.free(inc_snapshot_filename);
+
+                const inc_snapshot_url = try std.fmt.allocPrintZ(allocator, "http://{s}/{s}", .{
+                    r[0][0..r[1]],
+                    inc_snapshot_filename,
+                });
+                defer allocator.free(inc_snapshot_url);
+
+                logger.infof("downloading inc_snapshot from: {s}", .{inc_snapshot_url});
+                _ = downloadFile(
+                    allocator,
+                    logger,
+                    inc_snapshot_url,
+                    output_dir,
+                    inc_snapshot_filename,
+                    // NOTE: no min limit (we already downloaded the full snapshot at a good speed so this should be ok)
+                    null,
+                ) catch |err| {
+                    // failure here is ok (for now?)
+                    logger.infof("failed to download inc_snapshot: {s}", .{@errorName(err)});
+                    return;
+                };
+            }
+
+            // success
+            return;
+        }
+    }
+}
+
+const DownloadProgress = struct {
+    mmap: []align(std.mem.page_size) u8,
+    download_size: usize,
+    min_mb_per_second: ?usize,
+    logger: Logger,
+
+    mb_timer: std.time.Timer,
+    bytes_read: usize = 0,
+    file_memory_index: usize = 0,
+    has_checked_speed: bool = false,
+
+    const Self = @This();
+
+    pub fn init(
+        logger: Logger,
+        output_dir_str: []const u8,
+        filename: []const u8,
+        download_size: usize,
+        min_mb_per_second: ?usize,
+    ) !Self {
+        var output_dir = try std.fs.cwd().openDir(output_dir_str, .{});
+        var file = try output_dir.createFile(filename, .{ .read = true });
+        defer file.close();
+
+        // resize the file
+        try file.seekTo(download_size - 1);
+        _ = try file.write(&[_]u8{1});
+        try file.seekTo(0);
+
+        var file_memory = try std.os.mmap(
+            null,
+            download_size,
+            std.os.PROT.READ | std.os.PROT.WRITE,
+            std.os.MAP.SHARED,
+            file.handle,
+            0,
+        );
+
+        return .{
+            .logger = logger,
+            .mmap = file_memory,
+            .download_size = download_size,
+            .min_mb_per_second = min_mb_per_second,
+            .mb_timer = try std.time.Timer.start(),
+        };
+    }
+
+    pub fn bufferWriteCallback(ptr: [*c]c_char, size: c_uint, nmemb: c_uint, user_data: *anyopaque) callconv(.C) c_uint {
+        const len = size * nmemb;
+        var self: *Self = @alignCast(@ptrCast(user_data));
+        var typed_data: [*]u8 = @ptrCast(ptr);
+        var buf = typed_data[0..len];
+
+        @memcpy(self.mmap[self.file_memory_index..][0..len], buf);
```

Comment:
This looks a little weird to me. Seems fine, it's just not what I'm used to. Normally I would write it like this:
```suggestion
        @memcpy(self.mmap[self.file_memory_index .. self.file_memory_index + len], buf);
```
Do you know if it makes a difference? I'm not sure what the compiler optimizes away. My assumption is that it would be slightly faster to do the addition rather than slicing twice.

Reply:
not sure the difference but the double-slice approach seems to be the way stdlib does it